### ISSUE TITLE
Make `tracing_subscriber::span::{Data, Store, etc}` public

### ISSUE THREAD

**Issue by anp** [tokio-rs/tracing#505]
### Crates

tracing-subscriber

### Motivation

I'm implementing a `Subscriber` similar to `Formatter` in many ways and would like to reuse the span allocation it makes use of in `span::Store`.

### Proposal

Is it safe to expose these types as directly usable? That would probably be my preference to avoid having to implement my own slab/locking/etc setup.

### Alternatives

Is it possible to imagine a version of the crate where all I/O is done in layers separately from the span and metadata management?

Maybe Subscriber could be blanket impl'd for structs that implement some smaller trait, named something like `Storage`?

**Comment by davidbarsky** [comment]
@anp If I'm not mistaken, the [`LookupSpan`](https://tracing.rs/tracing_subscriber/registry/trait.lookupspan) trait + [`Registry`](https://tracing.rs/tracing_subscriber/struct.registry) struct should address your issue. However, you'd implement a _Layer_ instead of Subscriber. Here's a basic one done by @tobz: https://github.com/metrics-rs/metrics-tracing-test/blob/master/src/layer.rs.

**Comment by anp** [comment]
Huzzah! I think this is perfect actually, I would like all of my code to live in layers and to have the subscriber just act as the registry.

I'm tempted to rephrase this as a docs/tutorial request but I'm not well-versed enough in the subscriber crate yet to know where would be the best place.

**Comment by davidbarsky** [comment]
@anp you're right, this _does_ need to be documented and it's _not_ yet documented. it really does.
